feat(js): add wrapper to build pipeline

genaris opened the pull request:
Adds JavaScript wrapper to CI/CD pipeline, in a very much similar fashion to Indy VDR. Includes some minor fixes raised when running typing/lint checks.

It needs NPM_TOKEN to be added to the project config (I'm currently testing in my [fork](https://github.com/genaris/aries-askar/tree/feat/test-release) using test NPM package names).

Build pipeline was updated to match the one in indy-vdr, except of android/ios which is still handled in separate files (to be updated in #67) and Python wrapper for linux-aarch64 as it's throwing some errors that need further investigation.

When installed, Node.Js wrapper will retrieve, from Askar releases, the proper aries_askar binary. To determine the correct binary, it uses a script that provides the architecture to `node-pre-gyp` install script. This logic could be useful to indy-vdr-nodejs as well, as I've seen that it is not getting the correct file in all cases.

TimoGlastra reviewed (commented):
This is great @genaris! Some small remarks, mostly related to eslint ignores 😄

TimoGlastra commented:
> This logic could be useful to indy-vdr-nodejs as well, as I've seen that it is not getting the correct file in all cases.

I haven't noticed this myself. What issues have you encountered?

TimoGlastra commented:
@ryjones can you add the NPM_TOKEN secret? Thank you!

genaris commented on wrappers/javascript/aries-askar-nodejs/babel.config.js:
```diff
@@ -1,6 +1,6 @@
 const path = require('path')
 
-const ariesAskarShared = require('../shared/package.json')
+const ariesAskarShared = require('../aries-askar-shared/package.json')
```
I think it is needed for current Jest configuration. Maybe with some tweaks like in aries-framework-javascript setup it wouldn't be necessary anymore.

blu3beri replied:
I believe, but it might have been revised a couple of times, it was for importing from `aries-askar-shared` with hot reloading. Don't think it is still required, as we use import resolved. But this stuff always confuses me as we can do this in 100 different ways.

genaris replied:
OK, I've updated jest config to not need babel config file for NodeJS wrapper. I left it in react-native as I wasn't sure (had some errors when checking types) and will be updated in another PR anyway.

genaris commented on wrappers/javascript/aries-askar-react-native/babel.config.js:
```diff
@@ -0,0 +1,18 @@
+
+const ariesAskarShared = require('../aries-askar-shared/package.json')
+
+module.exports = {
```
Not sure in react-native world but I leave @blu3beri to give us some words about it :stuck_out_tongue:

blu3beri replied:
Same as above. :)

genaris commented on wrappers/javascript/aries-askar-react-native/src/index.ts:
```diff
@@ -1,3 +1,6 @@
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
```
I added them because I was having some errors and wanted to match the index from indy-vdr-react-native. Seems to not be needed so I'm reverting to original code.

genaris commented on wrappers/javascript/aries-askar-shared/src/crypto/Jwk.ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/no-cycle */
```
I fixed this dependency cycle by removing the method `Jwk.toKey`, as the same behaviour can be achieved by using `Key.fromJwk`. Again I leave the father of the creature (@blu3beri) to tell if it's reasonable or if we should do it in another way.

blu3beri replied:
Yeah, can be removed if we can achieve the same functionality :). I think I was just copying everything from the Python wrapper.

genaris commented on wrappers/javascript/lerna.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "packages": ["packages/*"],
```
Yes, this comes from a copy-paste from indy-vdr. That setting should not have any impact anyway, as we are setting `useWorkspaces`.

genaris commented on wrappers/javascript/package.json:
```diff
@@ -39,11 +37,14 @@
     "eslint-plugin-import": "^2.25.4",
     "eslint-plugin-prettier": "^4.0.0",
+    "lerna": "^6.4.1",
+    "metro-react-native-babel-preset": "^0.74.0",
```
It seems to be not needed so I've removed it.

genaris commented:
> > This logic could be useful to indy-vdr-nodejs as well, as I've seen that it is not getting the correct file in all cases.
> 
> I haven't noticed this myself. What issues have you encountered?

As of current state, node-pre-gyp expects a package name with the format `library-{platform}.tar.gz`, which for instance for a Linux box based on an x86-64 CPU resolves to the following URL:

> https://github.com/hyperledger/indy-vdr/releases/download/v0.4.0-dev.1/library-linux.tar.gz

This gives a 404, because the release artifacts do include machine architecture in their names (i.e. the correct URL for this case is https://github.com/hyperledger/indy-vdr/releases/download/v0.4.0-dev.1/library-linux-x86_64.tar.gz).

So I updated the expected naming to the format `library-{platform}-{arch}.tar.gz` and, by a helper arch.js script,  converted that **arch** to the naming used in the CI/CD scripts (because Node.JS os.arch() uses a different nomenclature and also in some cases like MacOS we are generating an universal bundle).

blu3beri commented:
> > > This logic could be useful to indy-vdr-nodejs as well, as I've seen that it is not getting the correct file in all cases.
> > 
> > 
> > I haven't noticed this myself. What issues have you encountered?
> 
> As of current state, node-pre-gyp expects a package name with the format `library-{platform}.tar.gz`, which for instance for a Linux box based on an x86-64 CPU resolves to the following URL:
> 
> > https://github.com/hyperledger/indy-vdr/releases/download/v0.4.0-dev.1/library-linux.tar.gz
> 
> This gives a 404, because the release artifacts do include machine architecture in their names (i.e. the correct URL for this case is https://github.com/hyperledger/indy-vdr/releases/download/v0.4.0-dev.1/library-linux-x86_64.tar.gz).
> 
> So I updated the expected naming to the format `library-{platform}-{arch}.tar.gz` and, by a helper arch.js script, converted that **arch** to the naming used in the CI/CD scripts (because Node.JS os.arch() uses a different nomenclature and also in some cases like MacOS we are generating an universal bundle).

I don't think we have to generate a universal bundle. My preference is just separating aarch64 and x86_64, I don't think anyone needs both at the same time (what a setup would that be). Also, it will keep the size slightly smaller.

genaris commented:
> I don't think we have to generate a universal bundle. My preference is just separating aarch64 and x86_64, I don't think anyone needs both at the same time (what a setup would that be). Also, it will keep the size slightly smaller.

Yes, it's a common practice in Mac OS world since the days of PowerPC/x86 but I agree that in this case it will be fine to publish separate artifacts for each architecture.

Anyway, the script might be useful to select the correct filename because `os.arch()` in NodeJS returns "x64" for x86_64 and "arm64" for aarch64. We can of course generate file names aligned to NodeJS nomenclature but I didn't want to change everything only due to this JS wrapper requirement.

blu3beri commented:
>  
> > I don't think we have to generate a universal bundle. My preference is just separating aarch64 and x86_64, I don't think anyone needs both at the same time (what a setup would that be). Also, it will keep the size slightly smaller.
> 
> Yes, it's a common practice in Mac OS world since the days of PowerPC/x86 but I agree that in this case it will be fine to publish separate artifacts for each architecture.
> 
> Anyway, the script might be useful to select the correct filename because `os.arch()` in NodeJS returns "x64" for x86_64 and "arm64" for aarch64. We can of course generate file names aligned to NodeJS nomenclature but I didn't want to change everything only due to this JS wrapper requirement.

Yeah, I see. I prefer the x86_64 / aarch64 naming, so if we can achieve that with that script, let's keep it for sure.

ryjones commented:
> @ryjones can you add the NPM_TOKEN secret? Thank you!

Done!

genaris commented:
> Is this ready to be merged @genaris?

Sure!
